nixos/modules/system/boot/loader/limine/limine.nix

```diff
@@ -0,0 +1,155 @@
+        It is recommended to set this to false, as it allows gaining root
+        access by passing `init=/bin/sh` as a kernel parameter.
+      '';
+    };
+
+    maxGenerations = mkOption {
+      default = 10;
+      example = 50;
+      type = types.nullOr types.int;
+      description = ''
+        Maximum number of latest generations in the boot menu.
+        Useful to prevent boot partition of running out of disk space.
+        `null` means no limit i.e. all generations that were not
+        garbage collected yet.
+      '';
+    };
+
+    extraEntries = mkOption {
+      default = { };
+      type = types.attrsOf types.str;
+      example = literalExpression ''
+        {
+          "memtest86" = '''
+            protocol: chainload
+            path: boot():///efi/memtest86/memtest86.efi
+          ''';
+        }
+      '';
+      description = ''
+        Any additional entries you want added to the Limine boot menu. Each attribute denotes
+        the display name of the boot entry, which need be formatted according to the Limine
+        documentation which you can find [here](https://github.com/limine-bootloader/limine/blob/trunk/CONFIG.md).
+      '';
+    };
+
+    additionalFiles = mkOption {
+      default = { };
+      type = types.attrsOf types.path;
+      example = literalExpression ''
+        { "efi/memtest86/memtest86.efi" = "''${pkgs.memtest86-efi}/BOOTX64.efi"; }
+      '';
+      description = ''
+        A set of files to be copied to {file}`/boot`. Each attribute name denotes the
+        destination file name in {file}`/boot`, while the corresponding attribute value
+        specifies the source file.
+      '';
+    };
+
+    useStorePaths = mkOption {
+      default = false;
+      type = types.bool;
+      description = ''
+        Use direct Nix store paths instead of copying boot files onto the boot partition. This is
+        noticeably slower but can greatly reduce the space usage on the boot partition.
+      '';
+    };
+
+    validateChecksums = mkOption {
+      default = true;
+      type = types.bool;
+      description = ''
+        Whether to validate file checksums before booting.
+      '';
+    };
+
+    panicOnChecksumMismatch = mkOption {
+      default = false;
+      type = types.bool;
+      description = ''
+        Whether or not checksum validation failure should be a fatal
+        error at boot time.
+      '';
+    };
+
+    package = mkOption {
+      default = pkgs.limine;
+      defaultText = literalExpression "pkgs.limine";
+      type = types.package;
+      description = ''
+        Which Limine package to use for installation.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = pkgs.stdenv.hostPlatform.isx86;
+        message = "Limine can only be installed on x86 platforms";
+      }
+      {
+        assertion = efi.canTouchEfiVariables;
+        message = "Limine can only be installed on EFI platforms";
```

Comment:
I'm fine with merging this without BIOS support *for now*, but it would be awesome to support it. `pkgs.limine` has the install command for BIOS and hybrid installations. It also has the enroll-config command which embeds the blake2b hash of the config file into the bootloader binary to make sure it isn't being tampered with - the bootloader EFI executable can then be verified by secure boot to get "secure boot" with limine.

Also, efi.canTouchEfiVariables is *not* a check for UEFI platforms, it should only affect whether EFI variables are modified during installation. It's entirely valid to install on a UEFI system without changing EFI variables (and configuring them manually)